include/osquery/http_client.h

```diff
@@ -0,0 +1,368 @@
+  std::shared_ptr<ssl_stream> ssl_sock_;
+  boost_system::error_code ec_;
+  bool ssl_connection = false;
+  static const long SHORT_READ_ERROR = 0x140000dbL;
+  /// Setting the default port to squid proxy default port
+  static const int PROXY_DEFAUT_PORT = 3128;
+  static const int HTTP_DEFAULT_PORT = 80;
+  static const int HTTPS_DEFAULT_PORT = 443;
+};
+
+template <typename T>
+class HTTP_Request : public T {
+ public:
+  struct Header {
+    Header(const std::string& name, const std::string& value)
+        : name_(name), value_(value) {}
+
+   private:
+    std::string name_;
+    std::string value_;
+    friend class HTTP_Request<T>;
+  };
+
+ public:
+  HTTP_Request(const std::string& url) : uri_(url) {}
+
+  boost::optional<std::string> remoteHost() {
+    return uri_.host().size() ? uri_.host() : boost::optional<std::string>();
+  }
+
+  boost::optional<std::string> remotePort() {
+    return uri_.port().size() ? uri_.port() : boost::optional<std::string>();
+  }
+
+  boost::optional<std::string> remotePath() {
+    return uri_.path().size() ? uri_.path() : boost::optional<std::string>();
+  }
+
+  boost::optional<std::string> protocol() {
+    return uri_.scheme().size() ? uri_.scheme()
+                                : boost::optional<std::string>();
+  }
+
+  HTTP_Request& operator<<(const Header& h) {
+    this->T::set(h.name_, h.value_);
+    return *this;
+  }
+
+  void uri(const std::string& url) {
+    uri_ = url;
+  }
+
+ private:
+  boost::network::uri::uri uri_;
+};
+
+template <typename T>
+class HTTP_Response : public T {
+ public:
+  HTTP_Response() {}
+  HTTP_Response(const T& resp) : T(resp) {}
+
+  template <typename ITER>
+  class Iterator;
+  class Header;
+
+  unsigned status() {
+    return this->T::result_int();
+  }
+
+  const std::string& body() {
+    return this->T::body;
+  }
+
+  Header headers() {
+    return Header(this);
+  }
+};
+
+template <typename T>
+template <typename ITER>
+class HTTP_Response<T>::Iterator {
+ public:
+  Iterator(ITER iter) : iter_(iter) {}
+
+  Iterator operator++() {
+    ++iter_;
+    return *this;
+  }
+
+  Iterator operator++(int) {
+    Iterator tmp = *this;
+    ++*this;
+    return tmp;
+  }
+
+  bool operator!=(const Iterator& it) const {
+    return (iter_ != it.iter_);
+  }
+
+  auto operator-> () {
+    return this;
+  }
+
+  std::string header_name() {
+    return std::string(iter_->name_string());
+  }
+
+  std::string header_value() {
+    return std::string(iter_->value());
+  }
+
+ private:
+  ITER iter_;
+};
+
+template <typename T>
+class HTTP_Response<T>::Header {
+ public:
+  Header(T* resp) : resp_(resp) {}
+
+  std::string operator[](const std::string& name) {
+    return std::string(resp_->T::operator[](name));
+  }
+
+  auto begin() {
+    return Iterator<decltype(resp_->T::begin())>(resp_->T::begin());
+  }
+
+  auto end() {
+    return Iterator<decltype(resp_->T::end())>(resp_->T::end());
+  }
+
+ private:
+  T* resp_;
```

Comment:
Could this be a `std::unique_ptr`?

Reply:
No, since I am not handing over the ownership.

Reply:
I'm not sure why this class even exists. As far as I can tell it is to make the client calling code see `std::string` instead of `string_view`. But this comes at significant run-time cost.